### PR TITLE
Tasking system for #12

### PR DESCRIPTION
**Pull request by Matslom**
I work on that for a long time with long breaks, I hope that I didn't miss anything.

Any thoughts about functionality?

### CODE REVIEW

File: app/Http/Controllers/Admin/Tools/TasksController.php [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        if (!$task) {
+            throw new TaskNotFoundException;
+        }
+
+        $time = explode(' ', $task->frequency);
+        // Explode month [3] and day of week [4] for select inputs
+        $time[3] = explode(',', $time[3]);
+        $time[4] = explode(',', $time[4]);
+
+
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Tools/TasksController.php]: Can you please remove this empty line?

File: app/Http/Controllers/Admin/Tools/TasksController.php [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    /**
+     * @param CreateOrUpdateTaskRequest $request
+     * @return \Illuminate\Routing\Redirector|\Illuminate\Http\RedirectResponse
+     */
+    public function save(CreateOrUpdateTaskRequest $request)
+    {
+        $inputs = $request->except('_csrf');
+        // Format time inputs to cron expression
+        $cron_format = $this->timeInputsToCronExpression($request->get('time'));
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Tools/TasksController.php]: Can you please rename this to `$cronFormat`?

File: app/Http/Controllers/Admin/Tools/TasksController.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    public function update(int $id, CreateOrUpdateTaskRequest $request)
+    {
+        $inputs = $request->except('_csrf');
+        $task = $this->tasksRepository->find($inputs['id']);
+        if (!$task) {
+            throw new TaskNotFoundException;
+        }
+
+        // Format time inputs to cron expression
+        $cron_format = $this->timeInputsToCronExpression($request->get('time'));
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Tools/TasksController.php]: Can you please rename this to `$cronFormat`?

File: app/Http/Controllers/Admin/Tools/TasksController.php [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        // Calculate next run date
+        $dt = new \DateTime();
+        $cron = CronExpression::factory($cron_format);
+        $next_run = $cron->getNextRunDate($dt->setTimestamp($task->last_run))->getTimestamp();
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Tools/TasksController.php]: Can you please rename this to `$nextRun`?

File: app/Http/Controllers/Admin/Tools/TasksController.php [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    /**
+     * Format time inputs to cron expression
+     *
+     * @param array $inputs
+     * @return string
+     */
+    private function timeInputsToCronExpression(array $inputs = []): string
+    {
+        $cron_format = '';
```
Comment (euantorano) [inline app/Http/Controllers/Admin/Tools/TasksController.php]: Can you please rename this to `$cronFormat`?

File: app/Http/Controllers/ScheduleManagerController.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @version   2.0.0
+ * @package   mybb/core
+ * @license   http://www.mybb.com/licenses/bsd3 BSD-3
+ */
+
+namespace MyBB\Core\Http\Controllers;
+
+class ScheduleManagerController extends AbstractController
+{
+
```
Comment (euantorano) [inline app/Http/Controllers/ScheduleManagerController.php]: Can you please remove this empty line?

File: app/Http/Requests/Tools/CreateOrUpdateTaskRequest.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * @license   http://www.mybb.com/licenses/bsd3 BSD-3
+ */
+
+namespace MyBB\Core\Http\Requests\Tools;
+
+use MyBB\Core\Http\Requests\AbstractRequest;
+
+class CreateOrUpdateTaskRequest extends AbstractRequest
+{
+
```
Comment (euantorano) [inline app/Http/Requests/Tools/CreateOrUpdateTaskRequest.php]: Can you please remove this empty line?

File: app/Tasking/ScheduleManager.php [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+use Illuminate\Contracts\Foundation\Application;
+use Illuminate\Contracts\Cache\Repository as CacheRepository;
+use MyBB\Settings\Store;
+use MyBB\Core\Database\Models\Task;
+use MyBB\Core\Exceptions\TaskFailedException;
+use MyBB\Core\Database\Repositories\TasksRepositoryInterface;
+
+class ScheduleManager
+{
+
```
Comment (euantorano) [inline app/Tasking/ScheduleManager.php]: Can you please remove this empty line?

File: database/migrations/2017_03_05_123720_create_tasks_table.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExit('tasks');
```
Comment (euantorano) [inline database/migrations/2017_03_05_123720_create_tasks_table.php]: This doesn't look quite right, shouldn't it be `Schema:: dropIfExists('tasks')`?

File: database/migrations/2017_03_08_160035_create_task_logs_table.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExit('task_logs');
```
Comment (euantorano) [inline database/migrations/2017_03_08_160035_create_task_logs_table.php]: This doesn't look quite right, shouldn't it be `Schema:: dropIfExists('tasks')`?

File: tests/unit/Database/Repositories/Eloquent/TasksRepositoryTest.php [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+class TasksRepositoryTest extends \PHPUnit_Framework_TestCase
+{
+    public function testCanBeConstructed()
+    {
+        $task = Mockery::mock('MyBB\Core\Database\Models\Task');
+        $taskLog = Mockery::mock('MyBB\Core\Database\Models\TaskLog');
+        $repository = new TasksRepository($task, $taskLog);
+
+        static::assertInstanceOf('MyBB\Core\Database\Repositories\Eloquent\TasksRepository', $repository);
```
Comment (euantorano) [inline tests/unit/Database/Repositories/Eloquent/TasksRepositoryTest.php]: Could you use `::class` here and above (and below - every class reference)?

File: app/Console/Commands/stubs/task.stub [hunk 12]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace DummyNamespace;
```
Comment (Shade-) [inline app/Console/Commands/stubs/task.stub]: Is this named like that on purpose?

Comment (Matslom) [inline app/Console/Commands/stubs/task.stub]: Yes, it's only template. [Laravel](https://github.com/laravel/framework/blob/5.4/src/Illuminate/Foundation/Console/stubs/request.stub) use this naming so I used too.

Comment (Shade-) [inline app/Console/Commands/stubs/task.stub]: Alright, just asking, I'm quite ground-level when talking about Laravel, still learning :)

**Comment by Matslom** [comment]
Sure. No problem

09.04.2017 15:51 "Filippo" <notifications@github.com> napisał(a):

File: app/Database/Models/TaskLog.php [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            $model->created_at = $model->freshTimestamp();
+        });
+    }
+
+    /**
+     * Get the task record associated with the log.
+     */
+    public function task()
+    {
+        return $this->hasOne(Task::class, 'id', 'task_id');
```
Comment (euantorano) [inline app/Database/Models/TaskLog.php]: @Matslom Any thoughts on this? I've not looked at the table structure for the `TaskLog` table, but I'd assume it has a `task_id` column? If so, this should be `belongsTo`.

Comment (Matslom) [inline app/Database/Models/TaskLog.php]: You are right. It's fixed now.

Comment (euantorano) [inline app/Database/Models/TaskLog.php]: Nice one, thanks. I think we're good to go on merging this then, right? If there's anything else needed it can always be added in another PR.

Comment (Matslom) [inline app/Database/Models/TaskLog.php]: Yes, you can merge if everything works.
